--- xinit-1.0.8/Makefile.am.orig	2008-03-06 23:08:48.000000000 +0100
+++ xinit-1.0.8/Makefile.am	2008-03-08 02:11:36.208958943 +0100
@@ -57,7 +57,7 @@
 	-DXINIT=@XINIT@
 
 CPP_FILES_FLAGS = \
-	-DXINITDIR=$(XINITDIR) $(PROGCPPDEFS) -DLIBDIR=$(libdir) \
+	-DXINITDIR=$(XINITDIR) $(PROGCPPDEFS) -DBINDIR=$(bindir) -DLIBDIR=$(libdir) \
 	-DSHELL_CMD=$(SHELL_CMD) $(STARTX_COOKIE_FLAGS) \
 	-D__libexecdir__="$(libexecdir)" \
 	-D__bindir__="$(bindir)"
--- xinit-1.0.8/startx.cpp	2005-11-15 05:03:10.000000000 +0100
+++ xinit-1.0.8/startx.cpp	2005-12-22 10:53:22.620963000 +0100
@@ -65,7 +65,7 @@
 userserverrc=$HOME/.xserverrc
 sysserverrc=XINITDIR/xserverrc
 defaultclient=XTERM
-defaultserver=XSERVER
+defaultserver=BINDIR/Xwrapper
 defaultclientargs=""
 defaultserverargs=""
 clientargs=""
--- xinit-1.0.8/xinit.c.orig	2008-03-06 23:08:48.000000000 +0100
+++ xinit-1.0.8/xinit.c	2008-03-08 02:12:06.010657245 +0100
@@ -166,6 +166,7 @@
 #define	OK_EXIT		0
 #define	ERR_EXIT	1
 
+static char *default_wrapper = BINDIR "/Xwrapper";
 static char *default_server = "X";
 static char *default_display = ":0";		/* choose most efficient */
 static char *default_client[] = {"xterm", "-geometry", "+1+1", "-n", "login", NULL};
@@ -374,7 +375,10 @@
 	if (argc == 0 ||
 #ifndef __UNIXOS2__
 	    (**argv != '/' && **argv != '.')) {
-		*sptr++ = default_server;
+		if (access(default_wrapper, X_OK) == 0)
+			*sptr++ = default_wrapper;
+		else
+			*sptr++ = default_server;
 #else
 	    (**argv != '/' && **argv != '\\' && **argv != '.' &&
 	     !(isalpha(**argv) && (*argv)[1]==':'))) {
