--- xinit-1.4.4/startx.in.orig	2025-03-09 21:44:26.000000000 +0100
+++ xinit-1.4.4/startx.in	2025-03-10 17:56:52.562463648 +0100
@@ -16,7 +16,6 @@ prefix="@prefix@"
 exec_prefix="@exec_prefix@"
 xinitdir="@XINITDIR@"
 xterm="@XTERM@"
-xserver="@XSERVER@"
 xinit="@XINIT@"
 bundle_id_prefix="@bundleidprefix@"
 xauth="@XAUTH@"
@@ -24,6 +23,7 @@ bindir="@bindir@"
 libexecdir="@libexecdir@"
 mk_cookie="@MCOOKIE@"
 has_cookie_maker="@HAS_COOKIE_MAKER@"
+xserver="${bindir}/Xwrapper"
 
 unset SESSION_MANAGER
 
diff -urNp -x '*.orig' xinit-1.4.1.org/xinit.c xinit-1.4.1/xinit.c
--- xinit-1.4.1.org/xinit.c	2019-03-03 19:58:11.000000000 +0000
+++ xinit-1.4.1/xinit.c	2022-04-06 18:27:19.734543460 +0000
@@ -90,6 +90,7 @@ char xserverrcbuf[256];
 #define TRUE 1
 #define FALSE 0
 
+static char *default_wrapper = BINDIR "/Xwrapper";
 static char *default_server = "X";
 static char *default_display = ":0";        /* choose most efficient */
 static char *default_client[] = {"xterm", "-geometry", "+1+1", "-n", "login", NULL};
@@ -189,7 +190,10 @@ main(int argc, char *argv[])
      */
     if (argc == 0 ||
         (**argv != '/' && **argv != '.')) {
-        *sptr++ = default_server;
+	if (access(default_wrapper, X_OK) == 0)
+	    *sptr++ = default_wrapper;
+	else
+            *sptr++ = default_server;
     } else {
         server_given = 1;
         *sptr++ = *argv++;
