diff -urNp -x '*.orig' xinit-1.4.1.org/Makefile.am xinit-1.4.1/Makefile.am
--- xinit-1.4.1.org/Makefile.am	2019-03-03 19:58:11.000000000 +0000
+++ xinit-1.4.1/Makefile.am	2022-04-06 18:27:19.734543460 +0000
@@ -49,7 +49,7 @@ PROGCPPDEFS = \
 	-DXINIT=@XINIT@
 
 CPP_FILES_FLAGS = \
-	-DXINITDIR=$(XINITDIR) $(PROGCPPDEFS) -DLIBDIR=$(libdir) \
+	-DXINITDIR=$(XINITDIR) $(PROGCPPDEFS) -DBINDIR=$(bindir) -DLIBDIR=$(libdir) \
 	-DSHELL_CMD=$(SHELL_CMD) $(STARTX_COOKIE_FLAGS) \
 	-D__libexecdir__="$(libexecdir)" \
 	-D__bindir__="$(bindir)" \
--- xinit-1.4.3/startx.cpp.orig	2025-01-11 14:43:08.074323254 +0100
+++ xinit-1.4.3/startx.cpp	2025-01-11 14:44:52.763756103 +0100
@@ -14,7 +14,7 @@ XCOMM
 
 xinitdir=XINITDIR
 xterm=XTERM
-xserver=XSERVER
+xserver=BINDIR/Xwrapper
 xinit=XINIT
 bundle_id_prefix=BUNDLE_ID_PREFIX
 xauth=XAUTH
diff -urNp -x '*.orig' xinit-1.4.1.org/xinit.c xinit-1.4.1/xinit.c
--- xinit-1.4.1.org/xinit.c	2019-03-03 19:58:11.000000000 +0000
+++ xinit-1.4.1/xinit.c	2022-04-06 18:27:19.734543460 +0000
@@ -90,6 +90,7 @@ char xserverrcbuf[256];
 #define TRUE 1
 #define FALSE 0
 
+static char *default_wrapper = BINDIR "/Xwrapper";
 static char *default_server = "X";
 static char *default_display = ":0";        /* choose most efficient */
 static char *default_client[] = {"xterm", "-geometry", "+1+1", "-n", "login", NULL};
@@ -189,7 +190,10 @@ main(int argc, char *argv[])
      */
     if (argc == 0 ||
         (**argv != '/' && **argv != '.')) {
-        *sptr++ = default_server;
+	if (access(default_wrapper, X_OK) == 0)
+	    *sptr++ = default_wrapper;
+	else
+            *sptr++ = default_server;
     } else {
         server_given = 1;
         *sptr++ = *argv++;
